File: backend/app/utils/sanitize.py
"""
ChatterMate - HTML Sanitization Utility
Copyright (C) 2024 ChatterMate

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>
"""

import re
import html
import urllib.parse
from typing import Optional


def decode_entities(text: str) -> str:
    """Decode HTML entities and URL encoding to catch obfuscated attacks"""
    # Decode HTML entities multiple times to catch nested encoding
    for _ in range(3):
        text = html.unescape(text)
    
    # Decode URL encoding multiple times
    for _ in range(3):
        try:
            text = urllib.parse.unquote(text)
        except:
            break
    
    return text


def sanitize_message(message: Optional[str]) -> Optional[str]:
    """
    Comprehensive sanitization to prevent XSS attacks including encoded payloads.
    
    Blocks:
    - All dangerous HTML tags (script, iframe, object, embed, svg, etc.)
    - JavaScript URIs (including encoded variants)
    - Data URIs (except data:image)
    - Event handlers (all on* attributes)
    - SVG-based XSS vectors
    - Encoded payloads (HTML entities, URL encoding, Unicode)
    - CSS expressions and behaviors
    - Base64 encoded attacks
    - Template literals and ES6 syntax
    
    Args:
        message: The message text to sanitize
        
    Returns:
        Sanitized message text or None if input is None
    """
    if not message:
        return message
    
    # First, decode any encoded content to catch obfuscated attacks
    decoded_message = decode_entities(message)
    
    # Check for encoded javascript: or data: URIs in decoded content
    dangerous_patterns = [
        r'javascript\s*:',
        r'data\s*:(?!image/)',
        r'vbscript\s*:',
        r'file\s*:',
        r'about\s*:',
    ]
    
    for pattern in dangerous_patterns:
        if re.search(pattern, decoded_message, re.IGNORECASE):
            # If found in decoded content, sanitize the original too
            message = re.sub(pattern, '', message, flags=re.IGNORECASE)
    
    # Remove all dangerous HTML tags (comprehensive list)
    dangerous_tags = [
        'script', 'iframe', 'frame', 'frameset', 'object', 'embed', 
        'applet', 'base', 'link', 'meta', 'style', 'svg', 'math',
        'form', 'input', 'button', 'textarea', 'select', 'option',
        'xml', 'xss', 'import', 'video', 'audio', 'track', 'source',
        'canvas', 'details', 'template', 'slot', 'noscript',
        'marquee', 'bgsound', 'keygen', 'command'
    ]
    
    for tag in dangerous_tags:
        # Remove opening and closing tags with any attributes
        message = re.sub(rf'<{tag}[^>]*>.*?</{tag}>', '', message, flags=re.IGNORECASE | re.DOTALL)
        message = re.sub(rf'<{tag}[^>]*/?>', '', message, flags=re.IGNORECASE)
    
    # Remove javascript: URIs (including encoded variants)
    message = re.sub(r'j\s*a\s*v\s*a\s*s\s*c\s*r\s*i\s*p\s*t\s*:', '', message, flags=re.IGNORECASE)
    message = re.sub(r'&#[x]?[0-9a-f]+;?', lambda m: decode_entities(m.group()), message)
    
    # Remove vbscript: URIs
    message = re.sub(r'vbscript:', '', message, flags=re.IGNORECASE)
    
    # Remove data: URIs that could contain HTML/JS (but preserve data:image)
    message = re.sub(r'data:(?!image/)[^,;]*[,;]', '', message, flags=re.IGNORECASE)
    
    # Remove expression() and behavior CSS properties
    message = re.sub(r'expression\s*\([^)]*\)', '', message, flags=re.IGNORECASE)
    message = re.sub(r'behavior\s*:[^;]*', '', message, flags=re.IGNORECASE)
    message = re.sub(r'-moz-binding\s*:', '', message, flags=re.IGNORECASE)
    
    # Remove import statements
    message = re.sub(r'@import', '', message, flags=re.IGNORECASE)
    
    # Remove all event handlers (comprehensive list)
    event_handlers = [
        'onabort', 'onactivate', 'onafterprint', 'onafterupdate', 'onbeforeactivate',
        'onbeforecopy', 'onbeforecut', 'onbeforedeactivate', 'onbeforeeditfocus',
        'onbeforepaste', 'onbeforeprint', 'onbeforeunload', 'onbeforeupdate', 'onblur',
        'onbounce', 'oncellchange', 'onchange', 'onclick', 'oncontextmenu', 'oncontrolselect',
        'oncopy', 'oncut', 'ondataavailable', 'ondatasetchanged', 'ondatasetcomplete',
        'ondblclick', 'ondeactivate', 'ondrag', 'ondragend', 'ondragenter', 'ondragleave',
        'ondragover', 'ondragstart', 'ondrop', 'onerror', 'onerrorupdate', 'onfilterchange',
        'onfinish', 'onfocus', 'onfocusin', 'onfocusout', 'onhelp', 'oninput', 'oninvalid',
        'onkeydown', 'onkeypress', 'onkeyup', 'onlayoutcomplete', 'onload', 'onlosecapture',
        'onmessage', 'onmousedown', 'onmouseenter', 'onmouseleave', 'onmousemove',
        'onmouseout', 'onmouseover', 'onmouseup', 'onmousewheel', 'onmove', 'onmoveend',
        'onmovestart', 'onpaste', 'onpropertychange', 'onreadystatechange', 'onreset',
        'onresize', 'onresizeend', 'onresizestart', 'onrowenter', 'onrowexit', 'onrowsdelete',
        'onrowsinserted', 'onscroll', 'onsearch', 'onselect', 'onselectionchange',
        'onselectstart', 'onstart', 'onstop', 'onsubmit', 'onunload',
        'oncanplay', 'oncanplaythrough', 'oncuechange', 'ondurationchange', 'onemptied',
        'onended', 'onloadeddata', 'onloadedmetadata', 'onloadstart', 'onpause', 'onplay',
        'onplaying', 'onprogress', 'onratechange', 'onseeked', 'onseeking', 'onstalled',
        'onsuspend', 'ontimeupdate', 'onvolumechange', 'onwaiting', 'ontoggle',
        'onauxclick', 'ongotpointercapture', 'onlostpointercapture', 'onpointercancel',
        'onpointerdown', 'onpointerenter', 'onpointerleave', 'onpointermove', 'onpointerout',
        'onpointerover', 'onpointerup', 'onwheel', 'onanimationcancel', 'onanimationend',
        'onanimationiteration', 'onanimationstart', 'ontransitioncancel', 'ontransitionend',
        'ontransitionrun', 'ontransitionstart', 'onshow', 'onsort', 'onpointerrawupdate'
    ]
    
    for handler in event_handlers:
        # Remove with quotes
        message = re.sub(rf'{handler}\s*=\s*["\'][^"\']*["\']', '', message, flags=re.IGNORECASE)
        # Remove without quotes
        message = re.sub(rf'{handler}\s*=\s*[^>\s]*', '', message, flags=re.IGNORECASE)
        # Remove with backticks (template literals)
        message = re.sub(rf'{handler}\s*=\s*`[^`]*`', '', message, flags=re.IGNORECASE)
    
    # Remove dangerous attributes
    dangerous_attrs = [
        'formaction', 'action', 'form', 'srcdoc', 'srcset', 'dynsrc', 'lowsrc',
        'ping', 'poster', 'background', 'code', 'codebase', 'archive', 'profile',
        'xmlns', 'xlink:href', 'attributename', 'from', 'to', 'values', 'begin',
        'autofocus', 'autoplay', 'controls', 'manifest', 'sandbox'
    ]
    
    for attr in dangerous_attrs:
        message = re.sub(rf'{attr}\s*=\s*["\'][^"\']*["\']', '', message, flags=re.IGNORECASE)
        message = re.sub(rf'{attr}\s*=\s*[^>\s]*', '', message, flags=re.IGNORECASE)
    
    # Remove HTML comments (can be used to hide malicious code)
    message = re.sub(r'<!--.*?-->', '', message, flags=re.DOTALL)
    
    # Remove CDATA sections
    message = re.sub(r'<!\[CDATA\[.*?\]\]>', '', message, flags=re.DOTALL | re.IGNORECASE)
    
    # Remove XML processing instructions
    message = re.sub(r'<\?.*?\?>', '', message, flags=re.DOTALL)
    
    # Remove null bytes and other control characters that can bypass filters
    message = re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]', '', message)
    
    # Remove Unicode directional override characters (can hide malicious code)
    message = re.sub(r'[\u202a-\u202e\u2066-\u2069]', '', message)
    
    return message

File: frontend/src/utils/sanitize.ts
/*
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import DOMPurify from 'dompurify'

// Register DOMPurify hooks once on module load to prevent duplicate registrations
DOMPurify.addHook('uponSanitizeElement', (node, data) => {
  // Block SVG elements entirely
  if (data.tagName === 'svg') {
    node.parentNode?.removeChild(node)
    return
  }

  // Block math elements
  if (data.tagName === 'math') {
    node.parentNode?.removeChild(node)
    return
  }

  // Block any foreign objects
  if (data.tagName === 'foreignobject') {
    node.parentNode?.removeChild(node)
    return
  }

  // SECURITY: Remove <img>, <area>, and <map> tags completely. Anchors are kept
  // (clickable links) but hardened in afterSanitizeAttributes: protocol-restricted
  // to http(s)/mailto and forced target="_blank" rel="noopener noreferrer nofollow".
  const element = node as HTMLElement
  const tagName = data.tagName?.toUpperCase()
  if (tagName === 'IMG' || tagName === 'AREA' || tagName === 'MAP') {
    element.parentNode?.removeChild(element)
  }
})

DOMPurify.addHook('afterSanitizeAttributes', (node) => {
  // Block javascript: URIs (including encoded variants)
  if (node.hasAttribute('href')) {
    const href = node.getAttribute('href') || ''
    try {
      const decodedHref = decodeURIComponent(href.toLowerCase())

      if (
        decodedHref.includes('javascript:') ||
        decodedHref.includes('data:text/html') ||
        decodedHref.includes('vbscript:') ||
        decodedHref.includes('about:') ||
        decodedHref.includes('file:')
      ) {
        node.removeAttribute('href')
      }
    } catch {
      // If decoding fails, use original value for comparison
      if (
        href.toLowerCase().includes('javascript:') ||
        href.toLowerCase().includes('data:text/html') ||
        href.toLowerCase().includes('vbscript:') ||
        href.toLowerCase().includes('about:') ||
        href.toLowerCase().includes('file:')
      ) {
        node.removeAttribute('href')
      }
    }
  }

  // Harden anchors so markdown links render clickable but safely: positively allow
  // only http(s)/mailto, and always open in a new tab with noopener.
  if (node.nodeName === 'A') {
    const href = (node.getAttribute('href') || '').trim()
    if (!/^(https?:|mailto:)/i.test(href)) {
      node.removeAttribute('href')
    } else {
      node.setAttribute('target', '_blank')
      node.setAttribute('rel', 'noopener noreferrer nofollow')
    }
  }

  // Block javascript: and malicious data: URIs in src
  if (node.hasAttribute('src')) {
    const src = node.getAttribute('src') || ''
    try {
      const decodedSrc = decodeURIComponent(src.toLowerCase())

      if (
        decodedSrc.includes('javascript:') ||
        decodedSrc.includes('data:text/html') ||
        decodedSrc.includes('vbscript:') ||
        decodedSrc.includes('about:') ||
        decodedSrc.includes('file:')
      ) {
        node.removeAttribute('src')
      }
    } catch {
      // If decoding fails, use original value for comparison
      if (
        src.toLowerCase().includes('javascript:') ||
        src.toLowerCase().includes('data:text/html') ||
        src.toLowerCase().includes('vbscript:') ||
        src.toLowerCase().includes('about:') ||
        src.toLowerCase().includes('file:')
      ) {
        node.removeAttribute('src')
      }
    }
  }

  // Check for CSS expressions in style attribute
  if (node.hasAttribute('style')) {
    const style = node.getAttribute('style') || ''
    try {
      const decodedStyle = decodeURIComponent(style.toLowerCase())

      if (
        decodedStyle.includes('expression(') ||
        decodedStyle.includes('behavior:') ||
        decodedStyle.includes('-moz-binding') ||
        decodedStyle.includes('import') ||
        decodedStyle.includes('javascript:') ||
        decodedStyle.includes('vbscript:')
      ) {
        node.removeAttribute('style')
      }
    } catch {
      // If decoding fails, use original value for comparison
      if (
        style.toLowerCase().includes('expression(') ||
        style.toLowerCase().includes('behavior:') ||
        style.toLowerCase().includes('-moz-binding') ||
        style.toLowerCase().includes('import') ||
        style.toLowerCase().includes('javascript:') ||
        style.toLowerCase().includes('vbscript:')
      ) {
        node.removeAttribute('style')
      }
    }
  }

  // Remove any attribute that starts with "on" (event handlers)
  Array.from(node.attributes).forEach((attr) => {
    if (attr.name.toLowerCase().startsWith('on')) {
      node.removeAttribute(attr.name)
    }
  })
})

/**
 * Comprehensive HTML sanitization to prevent all XSS attacks including:
 * - Encoded payloads (HTML entities, URL encoding, Unicode)
 * - SVG-based XSS vectors
 * - JavaScript URIs and protocol handlers
 * - Data URIs (except images)
 * - CSS expressions and injections
 * - Event handlers
 * - All dangerous HTML tags
 */
export function sanitizeHtml(html: string): string {
  // Configure DOMPurify with strict security settings
  const config = {
    // Block all dangerous tags including SVG, form elements and images.
    // NOTE: 'a' must NOT be in this list — FORBID_TAGS beats ALLOWED_TAGS, and
    // anchors are intentionally kept (markdown links) then hardened by the
    // afterSanitizeAttributes hook (http(s)/mailto only, forced target+rel).
    FORBID_TAGS: [
      'iframe', 'frame', 'frameset', 'object', 'embed', 'applet', 'script',
      'base', 'link', 'meta', 'style', 'svg', 'math', 'form', 'input',
      'button', 'textarea', 'select', 'option', 'xml', 'xss', 'import',
      'video', 'audio', 'track', 'source', 'canvas', 'details', 'template',
      'slot', 'noscript', 'marquee', 'bgsound', 'keygen', 'command',
      'img', 'area', 'map'  // SECURITY: Remove image/map tags completely
    ],

    // Block dangerous attributes
    FORBID_ATTR: [
      // Event handlers
      'onerror', 'onload', 'onclick', 'onmouseover', 'onmouseout', 'onmousemove',
      'onkeydown', 'onkeyup', 'onkeypress', 'onfocus', 'onblur', 'onchange',
      'onsubmit', 'ondblclick', 'oncontextmenu', 'oninput', 'oninvalid',
      'onreset', 'onsearch', 'onselect', 'onabort', 'oncanplay', 'oncanplaythrough',
      'oncuechange', 'ondurationchange', 'onemptied', 'onended', 'onloadeddata',
      'onloadedmetadata', 'onloadstart', 'onpause', 'onplay', 'onplaying',
      'onprogress', 'onratechange', 'onseeked', 'onseeking', 'onstalled',
      'onsuspend', 'ontimeupdate', 'onvolumechange', 'onwaiting', 'ontoggle',
      'onauxclick', 'ongotpointercapture', 'onlostpointercapture', 'onpointercancel',
      'onpointerdown', 'onpointerenter', 'onpointerleave', 'onpointermove',
      'onpointerout', 'onpointerover', 'onpointerup', 'onwheel', 'onanimationcancel',
      'onanimationend', 'onanimationiteration', 'onanimationstart', 'ontransitioncancel',
      'ontransitionend', 'ontransitionrun', 'ontransitionstart', 'ondrag', 'ondragend',
      'ondragenter', 'ondragleave', 'ondragover', 'ondragstart', 'ondrop', 'oncopy',
      'oncut', 'onpaste', 'onscroll', 'onmessage', 'onmouseenter', 'onmouseleave',
      'onmousewheel', 'onbeforeunload', 'onerrorupdate', 'onhelp', 'onmove',
      'onreadystatechange', 'onresize', 'onstart', 'onstop', 'onunload',
      'onactivate', 'onafterprint', 'onafterupdate', 'onbeforeactivate',
      'onbeforecopy', 'onbeforecut', 'onbeforedeactivate', 'onbeforeeditfocus',
      'onbeforepaste', 'onbeforeprint', 'onbeforeupdate', 'onbounce',
      'oncellchange', 'oncontrolselect', 'ondataavailable', 'ondatasetchanged',
      'ondatasetcomplete', 'ondeactivate', 'onfilterchange', 'onfinish',
      'onfocusin', 'onfocusout', 'onlayoutcomplete', 'onlosecapture',
      'onmoveend', 'onmovestart', 'onpropertychange', 'onresizeend',
      'onresizestart', 'onrowenter', 'onrowexit', 'onrowsdelete', 'onrowsinserted',
      'onselectionchange', 'onselectstart', 'onshow', 'onsort', 'onpointerrawupdate',

      // Dangerous attributes
      'formaction', 'action', 'form', 'srcdoc', 'srcset', 'dynsrc', 'lowsrc',
      'ping', 'poster', 'background', 'code', 'codebase', 'archive', 'profile',
      'xmlns', 'xlink:href', 'attributename', 'from', 'to', 'values', 'begin',
      'autofocus', 'autoplay', 'controls', 'manifest', 'sandbox',
      // SECURITY: block resource-loading attributes. 'href' is intentionally NOT
      // here (FORBID_ATTR beats ALLOWED_ATTR): markdown links need it, and the
      // afterSanitizeAttributes hook strips any href that isn't http(s)/mailto.
      'src', 'data'
    ],

    // Only allow safe protocols
    ALLOWED_URI_REGEXP: /^(?:(?:(?:f|ht)tps?|mailto|tel|callto|sms|cid|xmpp):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i,

    // SECURITY: Strip ALL HTML tags to prevent rendering exploits
    // Only allow basic text formatting for markdown (no links, images, or any potentially dangerous tags)
    ALLOWED_TAGS: [
      'a',
      'b', 'i', 'u', 'strong', 'em', 'p', 'br', 'ul', 'ol', 'li',
      'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'blockquote', 'code', 'pre',
      'table', 'thead', 'tbody', 'tr', 'th', 'td', 'span', 'div',
      'del', 'hr', 'sup', 'sub', 'abbr', 'cite', 'dfn', 'kbd', 'mark',
      'q', 'samp', 'small', 'time', 'var'
    ],

    // Allow safe link attributes only (href is protocol-restricted + target/rel are
    // forced by the afterSanitizeAttributes hook above). No src or resource-loading attrs.
    ALLOWED_ATTR: [
      'href', 'target', 'rel',
      'title', 'class', 'id', 'align', 'colspan', 'rowspan'
    ],

    // Return a string instead of a document
    RETURN_DOM: false,
    RETURN_DOM_FRAGMENT: false,

    // Keep HTML comments removed
    ALLOW_DATA_ATTR: false
    // NOTE: do NOT set USE_PROFILES here — it overrides ALLOWED_TAGS/ALLOWED_ATTR,
    // which would drop the <a> tags we explicitly allow above. The explicit
    // ALLOWED_TAGS/ALLOWED_ATTR allowlist is authoritative; protocols are still
    // restricted by the afterSanitizeAttributes hook (href → http/https/mailto only).
  }

  // Sanitize the HTML using the pre-configured hooks
  return DOMPurify.sanitize(html, config)
}
